### PR TITLE
Sequence unrolling with QM

### PR DESCRIPTION
**Pull request by stavros11**
Some updates in the QM driver for unrolling to work properly. In the existing implementation multiple acquisition variables are created for the same readout pulse (when unrolled, therefore copied multiple times) and this complicates the QUA program and increases execution time. This is fixed here.

Some benchmarks are still needed to confirm that the speed-up from unrolling is comparable to other instruments and to check what are the limitations in the sequence length or measurements per sequence.

Checklist:
- [ ] Reviewers confirm new code works as expected.
- [ ] Tests are passing.
- [ ] Coverage does not decrease.
- [ ] Documentation is updated.

### CODE REVIEW

**Review by alecandido — COMMENTED** [review]
It seems like I had a pending review since I forgot when...

File: src/qibolab/instruments/qm/acquisition.py
Comment (stavros11) [reply src/qibolab/instruments/qm/acquisition.py]: The only argument is to have a more readable error, but I understand that this can blow up the code so I removed in 555b12fa4c4dde3723b799211dc04de9fea94c4c.

That being said, I am not sure what is the correct approach when it comes to custom errors. For this case in particular, assume the following situation that can easily happen: the user provides a runcard without specifying neither threshold nor angle and asks for `DISCRIMINATION`. The driver will query `Qubit` for threshold and angle and get the default values which are `None` and `0` respectively. Therefore there will be no error in `__post_init__` (since cos and sin of `0` can be calculated) and the first error will appear in
https://github.com/qiboteam/qibolab/blob/6f8596621078cde059785f053ce2332f36c1116e/src/qibolab/instruments/qm/acquisition.py#L221
because `self.threshold is None`. If QUA provides a readable error there, fine, if no the user will have not idea what happened (ie. that he cannot ask for `DISCRIMINATION` if he does not provide thresholds in the runcard).

File: src/qibolab/instruments/qm/sequence.py
Comment (stavros11) [reply src/qibolab/instruments/qm/sequence.py]: https://github.com/qiboteam/qibolab/pull/738/commits/50af0a8f202c1e8666b806560989e33ffc5b24d6

File: src/qibolab/instruments/qm/sequence.py
Comment (stavros11) [reply src/qibolab/instruments/qm/sequence.py:74]: > Or is there any other reason?

No other reason, only deduplication. In principle this is not strictly required at the pulse level, because the waveforms are registered separately in the QUA config and you can have multiple pulses using the same waveform, but I think even better if we deduplicate even the pulse section. Also readout pulses are a bit more delicate because they are associated with variables (where we save acquisition) and these need to be deduplicated as well (so that we don't have multiple variables for the "same" readout when unrolling).

File: src/qibolab/instruments/qm/sequence.py
Comment (stavros11) [reply src/qibolab/instruments/qm/sequence.py:71]: Yes, I think this is more or less the proposal in https://github.com/qiboteam/qibolab/pull/792#issue-2108078322 part 1, point 1. I believe the ZI equivalent is `signal`: https://github.com/qiboteam/qibolab/blob/13bdcb4644c26dd48097de2a4dc0211349fce9fa/src/qibolab/instruments/zhinst.py#L149

File: src/qibolab/instruments/qm/controller.py
Comment (stavros11) [reply src/qibolab/instruments/qm/controller.py:206]: Thanks, that's interesting, I don't know much about it either. Here it was just moved here from another existing file that I dropped (simulator.py). Initially it was also just copied from a tutorial. It is only needed when using a cloud simulator provided by QM, not for the real instrument. Btw this simulator is not working for a while now, so we could even drop, I just left it in case it comes back in the future (because it is useful for some quick schedule tests).

**Comment by stavros11** [comment]
Thanks @alecandido. I addressed the review comments.

File: src/qibolab/instruments/qm/acquisition.py
Comment (alecandido) [reply src/qibolab/instruments/qm/acquisition.py]: > That being said, I am not sure what is the correct approach when it comes to custom errors.

I believe this case is similar to the relative imports we discussed with @hay-k: it depends.
In general, in Python the attempt is to duck type functions: until they work, they work. If they fail, it should be because some operation is impossible on the given value, then that operation is reported.

However, as you pointed out, this could generate deeply nested errors, when you were able to tell it would have caused an error much before.
My advice would be to often resist the temptation to catch errors too much in advance, because it will make your code quite tangled.

We could start to validate the user input, to catch some mismatch in user input immediately, and for this task `pydantic` would be really ideal. However, Qibolab has quite experienced users (like Qibocal), so I'd consider this a second-order problem for some more time.

File: src/qibolab/instruments/qm/sequence.py
Comment (alecandido) [reply src/qibolab/instruments/qm/sequence.py:74]: > No other reason, only deduplication. In principle this is not strictly required at the pulse level, because the waveforms are registered separately in the QUA config and you can have multiple pulses using the same waveform, but I think even better if we deduplicate even the pulse section whenever possible.

I wanted to understand because I still believe that deduplication should happen in the common layer, so it should not be a QM specific feature.
The goal would be to assign an ID per waveform at a higher level.

However, this is something for after 0.2, for the time being it is perfectly to have this feature here.

(I also wanted to understand which pulse attributes you were using to dedup)

Comment (alecandido) [reply src/qibolab/instruments/qm/sequence.py:74]: > Of course the exact string format I used is not relevant. Any kind of id (readable or no) that does not duplicate the `start`/`channel` and is of string type works.

I also don't care about the specific format.
For the sake of compression, you could `hash()` the attributes, but I'm pretty sure it's worth nothing.

File: src/qibolab/instruments/qm/sequence.py
Comment (alecandido) [reply src/qibolab/instruments/qm/sequence.py:71]: Definitely, it's just that the comment came before #792, and the related discussion :P
(but slightly before, we were just getting on the same page)

File: src/qibolab/instruments/qm/controller.py
Comment (alecandido) [reply src/qibolab/instruments/qm/controller.py:208]: I don't mind too much about keeping it, for the time being. It's already there and quite a small overhead.

It was more of a curiosity.
(I was also interested in whether we could hijack some data and use the simulator in the CI, without having access to the instrument. But if the simulator is not working anyhow is useless)

**Comment by alecandido** [comment]
> Thanks @alecandido. I addressed the review comments.

Just ask again for a review whenever you need to merge.

**Comment by stavros11** [comment]
> Just ask again for a review whenever you need to merge.

Please have another look, when you have time. I would like to merge this and #738 particularly because QM is currently connected to a single-qubit chip (not sure for how long) and people who are not core devs are using it, so it is better to not have them jump around branches. It also works as expected with the QPU, modulo some performance issues with unrolling. I was already aware of these and I am thinking of potential solutions but that's for another PR.

**Review by alecandido — APPROVED** [review]
Generally fine, I just left a few comments.

As usual, if it's urgent, keep going, there is nothing critical. Just open issues for everything you believe to be valuable and you have to leave behind

File: src/qibolab/instruments/qm/acquisition.py
Comment (stavros11) [reply src/qibolab/instruments/qm/acquisition.py]: eb853ecdfbb9b355af6a5c7f58714a155241306c

File: src/qibolab/instruments/qm/acquisition.py
Comment (stavros11) [reply src/qibolab/instruments/qm/acquisition.py]: eb853ecdfbb9b355af6a5c7f58714a155241306c

File: src/qibolab/instruments/qm/acquisition.py
Comment (stavros11) [reply src/qibolab/instruments/qm/acquisition.py]: eb853ecdfbb9b355af6a5c7f58714a155241306c

File: src/qibolab/instruments/qm/acquisition.py
Comment (stavros11) [reply src/qibolab/instruments/qm/acquisition.py]: d5b8b45738b829e76f26d798d91dbf75a51bc6d5

File: src/qibolab/result.py
Comment (alecandido) [inline src/qibolab/result.py:111]: I believe `int` could be platform dependent, and possibly is `np.int64`, despite usually being 32 bit for the C compiler. 
```suggestion
        self.samples: npt.NDArray[np.uint32] = np.array(data).astype(np.uint32)
```

It depends if it is interpreted as `np.intc` or `np.int_`.
https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.intc
https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.int_

Comment (alecandido) [reply src/qibolab/result.py:111]: It's `np.int_` (the `_` was much more likely than one letter more, it would have been quite arbitrary)
```py
[ins] In [4]: a = np.arange(10, dtype=int)
                                                                                                                                                                                                                                                                                                                                                                            
[ins] In [5]: a
Out[5]: array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
                                                                                                                                                                                                                                                                                                                                                                            
[ins] In [6]: a.dtype
Out[6]: dtype('int64')
```

File: src/qibolab/instruments/qm/acquisition.py
Comment (stavros11) [reply src/qibolab/instruments/qm/acquisition.py]: 27893f87d54baf03be93f1d54172609888e877ac, 2d0420e9b5f2e9fc734f77d69d77d46c5832faeb

File: src/qibolab/instruments/qm/acquisition.py
Comment (stavros11) [reply src/qibolab/instruments/qm/acquisition.py]: 0aaf828df50f48fa390d051bad0616b7c0f4d2bd

File: src/qibolab/instruments/qm/acquisition.py
Comment (stavros11) [reply src/qibolab/instruments/qm/acquisition.py:227]: I am not sure how to make it work with the loop, unless I create something like a `list`, but for this particular I prefer shadowing over the list.

File: src/qibolab/instruments/qm/acquisition.py
Comment (stavros11) [reply src/qibolab/instruments/qm/acquisition.py:271]: I think the structure is single level: each readout pulse is mapped to an `Acquisition` object. However multiple readout pulses may be mapped to the same `Acquisition`. 

We could implement it with a single dictionary `{ro_pulse: acquisition}`, I just think the current implementation which is basically the inverse of this dictionary, is more useful for the manipulations we do later on these objects. I guess it appears as nested because it is the inverse of a map that is non-invertible (in the math sense).

**Comment by stavros11** [comment]
@alecandido thanks for the second review. I will merge this to the `octaves` branch to reduce the number of active branches. I will soon merge all to main, but I would like to complete a few more tests with the QPU before doing so.

File: src/qibolab/instruments/qm/acquisition.py
Comment (alecandido) [reply src/qibolab/instruments/qm/acquisition.py:292]: That's perfectly fine. I was not aware of the problem. Thanks for the issue

**Pull request merged**
